test/test_transforms_v2_refactored.py

```diff
@@ -0,0 +1,326 @@
+    # This dictionary contains the number of rightmost dimensions that contain the actual data.
+    # Everything to the left is considered a batch dimension.
+    data_dims = {
+        datapoints.Image: 3,
+        datapoints.BoundingBox: 1,
+        # `Mask`'s are special in the sense that the data dimensions depend on the type of mask. For detection masks
+        # it is 3 `(*, N, H, W)`, but for segmentation masks it is 2 `(*, H, W)`. Since both a grouped under one
+        # type all kernels should also work without differentiating between the two. Thus, we go with 2 here as
+        # common ground.
+        datapoints.Mask: 2,
+        datapoints.Video: 4,
+    }.get(input_type)
+    if data_dims is None:
+        raise pytest.UsageError(
+            f"The number of data dimensions cannot be determined for input of type {input_type.__name__}."
+        ) from None
+    elif input.ndim <= data_dims or not all(input.shape[:-data_dims]):
+        # input is not batched or has a degenerate batch shape
+        return
+
+    batched_input = input.as_subclass(torch.Tensor)
+    batched_output = kernel(batched_input, *args, **kwargs)
+    actual = _unbatch(batched_output, data_dims=data_dims)
+
+    single_inputs = _unbatch(batched_input, data_dims=data_dims)
+    expected = tree_map(lambda single_input: kernel(single_input, *args, **kwargs), single_inputs)
+
+    assert_close(actual, expected, **tolerances)
+
+
+def check_kernel(
+    kernel,
+    input,
+    *args,
+    check_cuda_vs_cpu=True,
+    check_scripted_vs_eager=True,
+    check_batched_vs_single=True,
+    **kwargs,
+):
+    initial_input_version = input._version
+
+    output = kernel(input.as_subclass(torch.Tensor), *args, **kwargs)
+    # Most kernels just return a tensor, but some also return some additional metadata
+    if not isinstance(output, torch.Tensor):
+        output, *_ = output
+
+    # check that no inplace operation happened
+    assert input._version == initial_input_version
+
+    assert output.dtype == input.dtype
+    assert output.device == input.device
+
+    # TODO: we can improve performance here by not computing the regular output of the kernel over and over
+    if check_cuda_vs_cpu:
+        _check_kernel_cuda_vs_cpu(kernel, _to_tolerances(check_cuda_vs_cpu), input, *args, **kwargs)
+
+    if check_scripted_vs_eager:
+        _check_kernel_scripted_vs_eager(kernel, _to_tolerances(check_scripted_vs_eager), input, *args, **kwargs)
+
+    if check_batched_vs_single:
+        _check_kernel_batched_vs_single(kernel, _to_tolerances(check_batched_vs_single), input, *args, **kwargs)
+
+
+def _check_kernel_scripted_smoke(dispatcher, input, *args, **kwargs):
+    dispatcher_scripted = _script(dispatcher)
+    dispatcher_scripted(input.as_subclass(torch.Tensor), *args, **kwargs)
+
+
+def _check_dispatcher_dispatch_simple_tensor(dispatcher, kernel, input, *args, **kwargs):
+    if not isinstance(input, datapoints.Image):
+        return
```

Comment:
ah OK it's converted below. Why do we need this check at all though?

Reply:
Not sure I understand. We want to make sure that our dispatcher dispatches to the right kernel. There should be four possible branches:

- simple tensor
- datapoint
- PIL
- anything else (error)

Are you asking why we need the simple tensor dispatch test or why we need dispatch tests in general?

Reply:
Sorry I wasn't clear. I had missed a few things and you addressed the rest in

> I've opted to perform the check whether an input is valid inside the _check* function rather than in the overarching check to keep the latter clean. This is not only the case here, but also in the other functions. Open for discussion on this one though.

(No strong opinion on that for now)